pkg/webservice/handlers.go

```diff
@@ -1216,3 +1227,40 @@ func getStream(w http.ResponseWriter, r *http.Request) {
 		}
 	}
 }
+
+func checkHeader(h http.Header, key string, value string) bool {
```

Comment:
Should we return error if users use unsupported compression?

Reply:
That may be more clearer to users and I added it.

Reply:
This may not be necessary. Skipping it is a solution when dealing with unsupported compression types in requests. Do you think it's essential for users to require this one?

Reply:
Please take a look https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/406

It seems to me following the standard error can avoid the misunderstanding in the future.

Reply:
> It seems to me following the standard error can avoid the misunderstanding in the future.

*NO*. We should never return an error to the client if it request an encoding we don't understand. `Accept-Encoding: identity` is the default, which means the identity encoding is always allowed. Therefore, if an unacceptable encoding is requested, we simply send the request uncompressed and without a `Content-Encoding: gzip` header.

Reply:
> NO. We should never return an error to the client if it request an encoding we don't understand. Accept-Encoding: identity is the default, which means the identity encoding is always allowed. Therefore, if an unacceptable encoding is requested, we simply send the request uncompressed and without a Content-Encoding: gzip header.

That is an acceptable way to me, but I'd like to have more discussion for my own education :)

Should we support full representation of `Accept-Encoding` ( `weight` and `coding` )? If yes, we need to consider the `Accept-Encoding: gzip;q=1.0, identity; q=0`.

Or we ignore the `weight` and only check the existence of `gzip` from the `Accept-Encoding`. This is the solution adopted by this PR.

Reply:
There's a common saying in network protocol design: Be liberal in what you accept, strict in what you produce. In other words, we can get away with just checking for the substring `gzip` in `Accept-Encoding`, and produce *exactly* `Content-Encoding: gzip` in that case. If we choose not to compress due to size, or gzip was not requested, then we use the standard identity version. Weights are not really necessary; yes, they are part of the spec, but the client is only giving its preference; we do not have to honor it.

Reply:
I'd like to see a very simple function like: `GetCompressedWriter(headers, writer) (writer)` that checks for the gzip header and wraps the given writer with a gzip-compressed one, else returns the original writer. Then in any endpoint we want to (potentially compress), we just replace our writer with that one instead.

Reply:
That will cause a leak as the gzip writer must be closed for it not to leak. Calling close on the http.ResponseWriter is not possible so we need more. Probably the easiest solution is to use the same solution as we have for the loggingHandler(). We wrap the compression choice in a handler function, which then gets wrapped in the logging handler. That means we have it all in one place and expand on it with compressor pooling or other things in the future.

Example code, which is not complete but gives some idea on how we can close the compressor. That can be expanded to use a sync.Pool to not recreate the
```
type gzipResponseWriter struct {
	io.Writer
	http.ResponseWriter
}

func (w gzipResponseWriter) Write(b []byte) (int, error) {
	return w.Writer.Write(b)
}

func makeGzipHandler(fn http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if !strings.Contains(r.Header.Get("Accept-Encoding"), "gzip") {
			fn(w, r)
			return
		}
		w.Header().Set("Content-Encoding", "gzip")
                w.Header().Del("Content-Length")
		gz := gzip.NewWriter(w)
		defer gz.Close()
		gzr := gzipResponseWriter{Writer: gz, ResponseWriter: w}
		fn(gzr, r)
	}
}
```

Reply:
Thanks to everyone for all the advice. I have got some questions about what @wilfred-s said. 
In the provided example, the compress handler would be wrapped within the logging handler. So we first see which API is being called, and then decide whether to use the gzip handler, since we haven't decided to compress all APIs yet? Is there any misunderstanding?